File: lib/nodes/operators/join_one_to_one.dart
import '../node.dart';

class JoinOneToOne<T extends Map<String, dynamic>>
    extends Node2Input<List<T>, List<T>, List<T>> {
  /// Key of input1 items to match items from input2
  /// Can be a string or a Function(item of input1)
  final dynamic joinKey;

  /// Key of input1 items to populate with matches from input2
  /// Can be a string or a Function(item of input1, matching item of input2)
  final dynamic joinedKey;

  /// Key of input2 items to match with input1 joinKey
  /// Can be a string or a Function(item of input2)
  final dynamic matchingKey;

  JoinOneToOne(
    Node<List<T>> nodeI1,
    this.joinKey,
    Node<List<T>> nodeI2, [
    this.matchingKey = 'id',
    dynamic joinedKey,
  ])  : joinedKey = joinedKey ?? joinKey,
        super(nodeI1, nodeI2);

  @override
  Future<List<T>> process(
    List<T> input1,
    List<T> input2,
  ) async =>
      // We need to copy items, otherwise operators may modify parents operators items
      // TODO: real deep copy, this is a shallow copy
      List.of(input1).map((e) {
        final input1Element = Map<String, dynamic>.from(e) as T;
        final joinValue = _stringOrFunctionGet(input1Element, joinKey);
        if (joinValue != null)
          for (T input2Element in input2)
            if (_stringOrFunctionGet(input2Element, matchingKey) == joinValue) {
              _stringOrFunctionSet(input1Element, joinedKey, input2Element);
              break;
            }
        return input1Element;
      }).toList();

  _stringOrFunctionGet(Map o, dynamic getter) {
    if (getter is String)
      return o[getter];
    else if (getter is Function)
      return getter(o);
    else
      throw StateError("wrong getter $getter");
  }

  _stringOrFunctionSet(Map o, dynamic setter, dynamic v) {
    if (setter is String)
      return o[setter] = v;
    else if (setter is Function)
      return setter(o, v);
    else
      throw StateError("wrong getter $setter");
  }

  // _deepGet(dynamic o, String path) {
  //   final pathList = path.split('.');
  //   if (pathList.isEmpty || path == "") return o;

  //   if (o is List) {
  //     final intPath = int.tryParse(pathList[0]);
  //     if (intPath == null)
  //       throw StateError("You must provide an integer for list _deepGet");
  //     else
  //       return _deepGet(o[intPath], pathList.skip(1).join('.'));
  //   }

  //   if (o is Map) return _deepGet(o[pathList[0]], pathList.skip(1).join('.'));

  //   throw StateError("Can't _deepGet $o");
  // }

  // _deepSet(dynamic o, String path, dynamic v) {
  //   final pathList = path.split('.');
  //   if (pathList.isEmpty || path == "") return o;

  //   if (o is List) {
  //     final intPath = int.tryParse(pathList[0]);
  //     if (intPath == null)
  //       throw StateError("You must provide an integer for list _deepSet");
  //     else if (pathList.length == 1)
  //       return o.replaceRange(intPath, intPath + 1, [v]);
  //     else
  //       return _deepSet(o[intPath], pathList.skip(1).join('.'), v);
  //   }

  //   if (o is Map) {
  //     if (pathList.length == 1)
  //       return o[pathList[0]] = v;
  //     else
  //       return _deepSet(o[pathList[0]], pathList.skip(1).join('.'), v);
  //   }

  //   throw StateError("Can't _deepSet $o");
  // }

  // dynamic _deepCopy(dynamic o) {
  //   if (o is List)
  //     return List.generate(
  //       o.length,
  //       (index) => _deepCopy(o[index]),
  //     );

  //   if (o is Map)
  //     return Map.fromEntries(o.entries.map(
  //       (e) => MapEntry(
  //         e.key,
  //         _deepCopy(e.value),
  //       ),
  //     ));

  //   return o;
  // }
}

File: src/nodes/operators/join_one_to_one.ts
import _ from 'lodash'
import { Node } from '../node'
import { Node2Input } from '../node_with_input/node_2_input'

/**
 * Acts like an SQL one-to-one join.
 * 
 * First input is the data to populate, containing a join key.
 * Second input is the data used for population or joined table.
 */
export class JoinOneToOne<I1, I2>
    extends Node2Input<Array<I1>, Array<I2>, Array<I1>> {
    /**
     * Join key of first input's items.
     */
    readonly joinKey: string | ((item: I1) => string)

    /**
     * Key of second input's items to match with `joinKey`.
     */
    readonly matchingKey: string | ((item: I2) => string)

    /**
     * Key of first input's items to populate with matches from seconde input's items.
     */
    readonly joinedKey: string | ((item: I1, matching: I2) => void)

    constructor(nodeI1: Node<Array<I1>>,
        joinKey: string | ((item: I1) => any),
        nodeI2: Node<Array<I2>>,
        matchingKey: string | ((item: I2) => any),
        joinedKey?: string | ((item: I1, matching: I2) => any)
    ) {
        super(nodeI1, nodeI2)
        this.joinKey = joinKey
        this.matchingKey = matchingKey
        this.joinedKey = joinedKey || joinKey
    }

    async process(input1: Array<I1>, input2: Array<I2>): Promise<Array<I1>> {
        return _.cloneDeep(input1).map((input1Element) => {
            const joinValue = _.isString(this.joinKey)
                ? (input1Element as any)[this.joinKey]
                : (this.joinKey as ((item: I1) => any))(input1Element)
            if (!_.isNil(joinValue))
                for (var input2Element of input2) {
                    const matchingValue = _.isString(this.matchingKey)
                        ? (input2Element as any)[this.matchingKey]
                        : (this.matchingKey as ((item: I2) => any))(input2Element)
                    if (matchingValue === joinValue) {
                        if (_.isString(this.joinedKey))
                            (input1Element as any)[this.joinedKey] = input2Element
                        else
                            (this.joinedKey as ((item: I1, matching: I2) => void))(input1Element, input2Element)
                    }

                }
            return input1Element
        })
    }
}